### PR TITLE
Add MH all variable selector tests and fix for bug with it

### PR DESCRIPTION
**Pull request by gordoncaleb**
There's a possible error with the logProb calculation when the Metropolis Hastings variable selector selects more than a single variable. This fixes that and adds a tests for when more than a single variable is selected.

### CODE REVIEW

File: keanu-python/keanu/vertex/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
...
...uals", "GreaterThanOrEqual", "GreaterThan", "LessThanOrEqual", "LessThan", "NotEquals", "BooleanConcatenation", "BooleanReshape", "BooleanSlice", "BooleanTake", "Not", "Bernoulli", "CastToDouble", "ConstantDouble", "DoubleIf", "DoubleProxy", "Addition", "ArcTan2", "Difference", "Division", "MatrixMultiplication", "Max", "Min", "Multiplication", "Power", "Concatenation", "Abs", "ArcCos", "ArcSin", "ArcTan", "Ceil", "Cos", "Exp", "Floor", "LogGamma", "Log", "MatrixDeterminant", "MatrixInverse", "Permute", "Reshape", "Round", "Sigmoid", "Sin", "Slice", "Sum", "Take", "Tan", "Beta", "Cauchy", "ChiSquared", "Dirichlet", "Exponential", "Gamma", "Gaussian", "HalfCauchy", "HalfGaussian", "InverseGamma", "KDE", "Laplace", "LogNormal", "Logistic", "MultivariateGaussian", "Pareto", "SmoothUniform", "StudentT", "Triangular", "Uniform", "Print", "CastToInteger", "ConstantInteger", "IntegerIf", "IntegerProxy", "IntegerAddition", "IntegerDifference", "IntegerDivision", "IntegerMax", "IntegerMin", "IntegerMultiplication", "IntegerPower", "IntegerConcatenation", "IntegerAbs", "IntegerReshape", "IntegerSlice", "IntegerSum", "IntegerTake", "Binomial", "Multinomial", "Poisson", "UniformInt", "Assert"]
```
Comment (wesleyw72) [inline keanu-python/keanu/vertex/__init__.py]: Nice, looks like the permute was missing.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 2]
```diff
@@ -20,43 +24,51 @@
 import java.util.Collections;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNull;
+import static org.junit.Assert.*;
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: 👍

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 3]
```diff
@@ -93,61 +105,57 @@ public void samplesContinuousTensorPrior() {
...
             1000
         );
 
-        double postProbTrue = posteriorSamples.get(A).probability(v -> v.scalar());
+        testCase.assertExpected(posteriorSamples);
+    }
+
+    @Category(Slow.class)
+    @Test
+    public void samplesComplexDiscreteWithMultipleVariableSelect() {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Called it `samplesComplexDiscreteWithFullVariableSelect`

File: keanu-python/keanu/vertex/__init__.py (hunk 1 above)
Comment (gordoncaleb) [reply keanu-python/keanu/vertex/__init__.py]: 👍

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 4]
```diff
@@ -93,61 +105,57 @@ public void samplesContinuousTensorPrior() {
...
-            .orElse(0.1);
+    public void samplesSimpleDiscretePriorWithDefaults() {
 
-        BernoulliVertex C = new BernoulliVertex(B);
+        MCMCTestCase testCase = new SingleVariateDiscreteTestCase();
 
-        C.observe(true);
-
-        BayesianNetwork bayesNet = new BayesianNetwork(Arrays.asList(A, B, C));
-        bayesNet.probeForNonZeroProbability(100);
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Both of the test cases have non-zero probability for the entire space, which means you will never need to probe for non-zero probability.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Called it `samplesComplexDiscreteWithFullVariableSelect`

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 5]
```diff
@@ -93,61 +105,57 @@ public void samplesContinuousTensorPrior() {
...
-
-        BernoulliVertex A = new BernoulliVertex(0.5);
-        BernoulliVertex B = new BernoulliVertex(0.5);
-
-        BooleanVertex C = A.or(B);
-
-        DoubleVertex D = If.isTrue(C)
-            .then(0.9)
-            .orElse(0.1);
+    public void samplesComplexDiscretePriorWithDefaults() {
```
[the remark this answers is not in the document]
Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: I'd like to make sure that given whatever defaults we choose that it works. If we change these defaults then we'd have to change this test. This test is explicitly meant to test our assumptions made in the defaults.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 6]
```diff
@@ -31,33 +35,47 @@
...
+    public void samplesContinuousPriorSingleVariableSelected() {
 
-        DoubleVertex A = new GaussianVertex(20.0, 1.0);
-        DoubleVertex B = new GaussianVertex(20.0, 1.0);
+        MCMCTestCase testCase = new SumGaussianTestCase();
 
-        A.setValue(20.0);
-        B.setValue(20.0);
+        BayesianNetwork bayesNet = testCase.getModel();
+        KeanuProbabilisticModel model = new KeanuProbabilisticModel(bayesNet);
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Do you need two lines here?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 7]
```diff
@@ -31,33 +35,47 @@
...
-        double averagePosteriorA = posteriorSamples.getDoubleTensorSamples(A).getAverages().scalar();
-        double averagePosteriorB = posteriorSamples.getDoubleTensorSamples(B).getAverages().scalar();
+        BayesianNetwork bayesNet = testCase.getModel();
+        KeanuProbabilisticModel model = new KeanuProbabilisticModel(bayesNet);
 
-        double actual = averagePosteriorA + averagePosteriorB;
-        assertEquals(44.0, actual, 0.1);
+        NetworkSamples posteriorSamples = MetropolisHastings.builder()
+            .proposalDistribution(new PriorProposalDistribution(model.getLatentVertices()))
+            .rejectionStrategy(new RollBackToCachedValuesOnRejection(model.getLatentVertices()))
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: Would it be helpful for Python users to be able to set the rejection strategy?  They can't do that at the moment.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java [hunk 8]
```diff
@@ -95,63 +113,62 @@ public void samplesContinuousTensorPrior() {
...
+            .variableSelector(MHStepVariableSelector.FULL_VARIABLE_SELECTOR)
+            .build()
+            .getPosteriorSamples(
+                model,
+                model.getLatentVertices(),
+                1000
+            );
```
Comment (christophernorth) [inline keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsTest.java]: There's quite a lot of fixed structure to a lot of these test cases now - could we turn it in to a single function that each test case calls?

**Pull request merged**